src/main.cpp

```diff
@@ -16,127 +18,157 @@
 #include "animations/Rainbow.h"
 #include "animations/Stars.h"
 #include "animations/Stripes.h"
+#include "animations/Strobe.h"
 
 // Pins for the rotary
 uint8_t rotary1 = 2;
 uint8_t rotary2 = 3;
 
 int buttonPin = A0;
-bool offMode = false;
 
 CRGB leds[NUMPIXELS];
-int feedbackPattern = -1;
-MyKnob knob(rotary1, rotary2, offMode, feedbackPattern);
 
 // Load animations...
-Crossfade crossfade(knob, leds);
-ColorChooser color_chooser(knob, leds);
-Race race(knob, leds);
-Stars stars(knob, leds);
-Rainbow rainbow(knob, leds);
-FuckMyEyes fuck_my_eyes(knob, leds);
-Stripes stripes(knob, leds);
-DiamondNecklace diamond_necklace(knob, leds);
-Dimmer dimmer(knob, leds);
-
-int animation_index = 0;
-Radio radio(knob, animation_index);
-
-// Animation *current_animation = &rainbow;
-Animation *current_animation = &crossfade;
-int previous_animation_index = -1;
-
-int feedbackLength = 800;
-long feedbackEnd = (millis() + feedbackLength);
-
-void runAdjustments() {
-  if (feedbackPattern < 0) {
-    return;
+Crossfade crossfade(leds);
+ColorChooser color_chooser(leds);
+Race race(leds);
+Stars stars(leds);
+Rainbow rainbow(leds);
+FuckMyEyes fuck_my_eyes(leds);
+Stripes stripes(leds);
+DiamondNecklace diamond_necklace(leds);
+Dimmer dimmer(leds);
+Strobe strobe(leds);
+
+const int NUM_ANIMATONS = 9;
+// const int NUM_ANIMATONS = 2;
+Animation* animations[NUM_ANIMATONS] = {
+    // &crossfade, &color_chooser};
+    &crossfade, &color_chooser,    &race,  &stars, &rainbow, &fuck_my_eyes,
+    &stripes,   &diamond_necklace, &dimmer};
+
+int old_animation_index = 0;
+MyKnob knob{};
+Radio radio(knob, old_animation_index);
+
+class BaseController {
+ private:
+   bool activeFlag = false;
+
+ public:
+  void setAsActive() { activeFlag = true; }
+
+  virtual void run() {
```

Comment:
What do these virtual functions get us in this app?

Reply:
Virtual methods go on base classes and are likely to be overridden in derived classes. This helps make sure that the derived class's method gets called when what you have it a pointer to an instance of a derived class but the pointer itself has the type of the base class.  A better explanation is [here](https://www.geeksforgeeks.org/virtual-function-cpp/).

Basically, I use virtual if I know I'm going to override that method in a derived class. By always using virtual it acts like Java does by default, which I know a lot better than C++.

Reply:
Good explanation! Thanks!